libtransmission/tr-assert.mm

```diff
@@ -3,6 +3,8 @@
 // or any future license endorsed by Mnemosyne LLC.
 // License text can be found in the licenses/ folder.
 
+#if defined(__clang__)
```

Comment:
Maybe better to change CMakeLists to include these files only when target is macOS/APPLE?

Reply:
It's already the case: but apparently @barracuda156 case was using GCC on APPLE. And we don't want those files compiled with GCC.

Reply:
GCC does support Objective C/C++, so perhaps it is worth testing whether files in question build or not.
What GCC does not support is the specific Clang-only flag.

Reply:
By disabling this code under gcc, won't the app just fail to compile on macOS + gcc?

If so, it might be better to leave these .mm files alone and have CMake error out with a more explicit error instead, e.g. `message(FATAL_ERROR, 'macOS builds require clang because gcc does not support arc yet')` or something similar.

Reply:
It should not require Clang, of course, not at least when built with GTK.

Reply:
I may be missing something but it looks like this PR disables `tr_strv_convert_utf8()` when building with GCC, and that function is used in libtransmission by `torrent-metainfo.cc` when loading torrents. So this seems pretty central to the app?

Reply:
Well, please do not merge this yet, my point in https://github.com/transmission/transmission/issues/5626 was that flags are wrong, not that files should not be built.

Reply:
> Well, please do not merge this yet

OK. Marking this PR as a draft to indicate it's not ready for merging.

Reply:
@ckerr Thank you.
I will update once something is clear, currently this is the stopper: https://github.com/transmission/transmission/issues/5627

Reply:
May we should just build `.cc` versions of both instead of `.mm` (conditional on compiler)?

Reply:
@ckerr @Coeur Yes, it does. Took a while, I had to rebuild a few times.

So provisionally what we need is, for now:

– remove clangish flags for gcc;
– fix macro in `utils.cc` for Apple + gcc – to have `tr_strv_convert_utf8()`;
– exclude those two `.mm` files for gcc build, but include `tr-assert.cc` (i.e. invert what is there in CMakeLists now, conditional on compiler used).

Older macOS will also need `-D_GLIBCXX_USE_CXX11_ABI=0`, otherwise linking fails spectacularly :)

Reply:
@nevack I have arrived to the same conclusion in fact, to use CMakeLists. But we can still get `tr_strv_convert_utf8()` by fixing a macro, so that it is not hidden for gcc build on Apple.

Reply:
@Coeur Yeah, let me just find a neat way to put that into CMakeLists. (In Macports I can make patches conditional on compiler or build_arch, which makes life easier.)

Maybe you could suggest how to rewrite this chunk?
https://github.com/transmission/transmission/blob/ddac05954b046b55f4bf25f45204f83f8ad8da4b/libtransmission/CMakeLists.txt#L207-L211
What I need is do exactly reverse when GCC is used on Apple.

Reply:
I was hoping you'd figured that part out, ah ah.
Well, with my cmake limited talent, I would write `tr_allow_compile_if(` three times:
1. first case is for everything except tr-assert/utils
2. second case is encapsulated in `if(CMAKE_CXX_COMPILER_ID MATCHES "Clang")` and keeps lines 207-211 as they are.
3. third case is the `else` and do the opposite of lines 207-211.

Reply:
I pushed a commit to show my idea (untested)

Reply:
> I was hoping you'd figured that part out, ah ah.

Ha-ha. It kinda uses a weird syntax which is unfamiliar to me. (But after all, I can dig into documentation.)

But you have mostly done everything here. What has to be changed in addition is this:
https://github.com/transmission/transmission/blob/ddac05954b046b55f4bf25f45204f83f8ad8da4b/libtransmission/utils.cc#L290
Should be `#if !(defined(__APPLE__) && defined(__clang__))` or something to this effect.

P. S. macosx/CMakeLists.txt also has `-fobjc-arc`, perhaps worth applying identical fix there.

Reply:
The whole macosx directory is reserved for clang compiler.

Reply:
ready for review

Reply:
@Coeur Thank you for working on this!